pkg/clusteragent/autoscaling/workload/external/tls.go

```diff
@@ -0,0 +1,225 @@
+
+package external
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+)
+
+// TLSConfig represents the TLS configuration used for external recommender calls.
+type TLSConfig struct {
+	CAFile             string
+	CertFile           string
+	KeyFile            string
+	ServerName         string
+	MinVersion         string
+	MaxVersion         string
+	InsecureSkipVerify bool
+}
+
+func (c *TLSConfig) requiresClientCertificate() bool {
+	return c != nil && (c.CertFile != "" || c.KeyFile != "")
+}
+
+const (
+	certificateCacheTimeout           = 10 * time.Minute
+	certificateCacheExpirationTimeout = 10 * time.Minute
+	certificateErrorCacheTimeout      = 1 * time.Minute
+)
+
+var tlsVersions = map[string]uint16{
+	"TLS13": tls.VersionTLS13,
+	"TLS12": tls.VersionTLS12,
+	"TLS11": tls.VersionTLS11,
+	"TLS10": tls.VersionTLS10,
+}
+
+// configureTransportTLS updates the provided transport with a TLS configuration based on the given settings.
+func configureTransportTLS(transport *http.Transport, config *TLSConfig, cache *tlsCertificateCache) error {
+	if config == nil {
+		return nil
+	}
+
+	tlsConfig, err := buildTLSConfig(config)
+	if err != nil {
+		return fmt.Errorf("failed to build TLS config: %w", err)
+	}
+
+	if config.requiresClientCertificate() {
+		if cache == nil {
+			return errors.New("tls certificate cache is required when using client certificates")
+		}
+		if config.CertFile == "" || config.KeyFile == "" {
+			return errors.New("both cert file and key file must be provided for client TLS configuration")
+		}
+		tlsConfig.GetClientCertificate = cache.GetClientCertificateReloadingFunc(config.CertFile, config.KeyFile)
+	}
+
+	transport.TLSClientConfig = tlsConfig
+	return nil
+}
+
+func buildTLSConfig(config *TLSConfig) (*tls.Config, error) {
+	var minVersion uint16
+	if config.MinVersion != "" {
+		var ok bool
+		minVersion, ok = tlsVersions[config.MinVersion]
+		if !ok {
+			return nil, fmt.Errorf("unknown minimum TLS version: %s", config.MinVersion)
+		}
+	}
+
+	var maxVersion uint16
+	if config.MaxVersion != "" {
+		var ok bool
+		maxVersion, ok = tlsVersions[config.MaxVersion]
+		if !ok {
+			return nil, fmt.Errorf("unknown maximum TLS version: %s", config.MaxVersion)
+		}
+	}
+
+	var rootCA *x509.CertPool
+	if config.CAFile != "" {
+		caPEM, err := os.ReadFile(config.CAFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read CA file: %w", err)
+		}
+		rootCA = x509.NewCertPool()
+		if !rootCA.AppendCertsFromPEM(caPEM) {
+			return nil, fmt.Errorf("failed to append root CA to pool")
+		}
+	}
+
+	return &tls.Config{
+		MinVersion:         minVersion,
+		MaxVersion:         maxVersion,
+		ServerName:         config.ServerName,
+		RootCAs:            rootCA,
+		InsecureSkipVerify: config.InsecureSkipVerify,
+	}, nil
+}
+
+// tlsCertificateCache is an expiring cache to store certificates in memory used for TLS connections.
+type tlsCertificateCache struct {
```

Comment:
nit: It's not really a cache if it owns the logic of reloading the items, it's more of a `tlsCertificateManager` or `Reloader` or something like that